go/marshal/encode.go

```diff
@@ -277,6 +282,10 @@ type encoderCacheT struct {
 
 var encoderCache = &encoderCacheT{}
 
+// Separate Set encoder cache because the same type with and without the
+// `noms:",set"` tag encode differently (Set vs Map).
+var setEncoderCache = &encoderCacheT{}
```

Comment:
The tuple would probably be (reflect.Type, types.NomsKind). I like the way it is now but if you prefer that I can change it.